std.getopt: Make private functions safe.

jondegenhardt opened the pull request:
This change make a pair of internal std.getopt functions `@safe`. This enables some additional `@safe` getopt use: callbacks via delegates and config option setting. `@safe` unit tests for both are included.

I also went through all the existing unit tests to see if any could be converted from `@system` to `@safe`. However, the existing `@system` unit tests take the address of local variables are cannot be `@safe`.

This PR is a follow-up to PR #5347.

JackStouffer commented:
`@nogc pure nothrow` should be added as well if applicable

quickfur commented:
And yes, if pure, nothrow, etc., are applicable, they should be added as well (to unittests and non-template functions only, of course).

jondegenhardt commented:
Added `pure nothrow @nogc` to `setConfig`. Also changed it to use `final switch`. However, `optMatch` calls functions not yet marked, so not none of these attributes could be applied to it. Also means none of the unit tests can be marked with these attributes.

CyberShadow commented:
This pull request introduced a regerssion:
https://issues.dlang.org/show_bug.cgi?id=17650
